File: llm_written/claude-sonnet-4.5/smith-waterman/python/test_smith_waterman.py
#!/usr/bin/env python3
"""
Differential test harness for Smith-Waterman algorithm.
Tests correctness (sequential vs parallel), determinism (repeated parallel runs), 
and performance on various input sizes.
"""

import sys
import time
import hashlib
import os
from typing import Tuple
from smith_waterman_baseline import SmithWaterman as BaselineSW
from algo_parallel import SmithWaterman as ParallelSW


def hash_result(result: Tuple[str, str, float, int]) -> str:
    """Create deterministic hash of alignment result."""
    aligned_a, aligned_b, score, identity = result
    content = f"{aligned_a}|{aligned_b}|{score}|{identity:.6f}"
    return hashlib.sha256(content.encode()).hexdigest()[:16]


def generate_sequence(length: int, seed: int = 42) -> str:
    """Generate deterministic DNA sequence."""
    import random
    random.seed(seed)
    bases = ['A', 'C', 'G', 'T']
    return ''.join(random.choice(bases) for _ in range(length))


def run_test_case(name: str, query: str, reference: str, workers: int = None):
    """Run differential test for a single case."""
    print(f"\n{'='*60}")
    print(f"Test Case: {name}")
    print(f"Query length: {len(query)}, Reference length: {len(reference)}")
    print(f"{'='*60}")
    
    # Baseline sequential
    baseline = BaselineSW(matchScore=2, mismatchScore=-1, gapScore=-1)
    
    t0 = time.perf_counter()
    result_seq = baseline.findAlignment(query, reference)
    t_seq = time.perf_counter() - t0
    
    hash_seq = hash_result(result_seq)
    print(f"✓ Sequential: {t_seq:.4f}s, hash={hash_seq}")
    
    # Parallel run 1
    parallel = ParallelSW(matchScore=2, mismatchScore=-1, gapScore=-1, workers=workers)
    
    t0 = time.perf_counter()
    result_par1 = parallel.findAlignment(query, reference)
    t_par1 = time.perf_counter() - t0
    
    hash_par1 = hash_result(result_par1)
    print(f"✓ Parallel-1: {t_par1:.4f}s, hash={hash_par1}")
    
    # Parallel run 2 (determinism check)
    t0 = time.perf_counter()
    result_par2 = parallel.findAlignment(query, reference)
    t_par2 = time.perf_counter() - t0
    
    hash_par2 = hash_result(result_par2)
    print(f"✓ Parallel-2: {t_par2:.4f}s, hash={hash_par2}")
    
    # Correctness check
    if result_seq != result_par1:
        print(f"❌ FAIL: Sequential and Parallel outputs differ!")
        print(f"   Sequential: {result_seq}")
        print(f"   Parallel:   {result_par1}")
        return False
    
    # Determinism check
    if hash_par1 != hash_par2:
        print(f"❌ FAIL: Parallel runs are non-deterministic!")
        print(f"   Run 1 hash: {hash_par1}")
        print(f"   Run 2 hash: {hash_par2}")
        return False
    
    print(f"✅ PASS: Correctness and determinism verified")
    
    # Performance summary (only for larger inputs)
    if len(query) * len(reference) >= 10000:
        speedup = t_seq / t_par1 if t_par1 > 0 else 0
        print(f"   Speedup: {speedup:.2f}x (seq={t_seq:.4f}s, par={t_par1:.4f}s)")
        return True, t_seq, t_par1, speedup
    
    return True, None, None, None


def main():
    """Run comprehensive differential test suite."""
    print("Smith-Waterman Differential Test Suite")
    print(f"Workers: {os.cpu_count()} cores detected")
    
    results = []
    
    # Edge cases
    print("\n" + "="*60)
    print("EDGE CASES")
    print("="*60)
    
    # Empty sequences
    success, *_ = run_test_case("Empty query", "", "ACGT", workers=2)
    results.append(("Edge: Empty query", success))
    
    success, *_ = run_test_case("Empty reference", "ACGT", "", workers=2)
    results.append(("Edge: Empty reference", success))
    
    success, *_ = run_test_case("Both empty", "", "", workers=2)
    results.append(("Edge: Both empty", success))
    
    # Single character
    success, *_ = run_test_case("Single char match", "A", "A", workers=2)
    results.append(("Edge: Single char", success))
    
    # Small identical
    success, *_ = run_test_case("Small identical", "ACGT", "ACGT", workers=2)
    results.append(("Edge: Small identical", success))
    
    # Small cases
    print("\n" + "="*60)
    print("SMALL CASES")
    print("="*60)
    
    success, *_ = run_test_case(
        "Small similar (10x10)", 
        generate_sequence(10, seed=1),
        generate_sequence(10, seed=2),
        workers=2
    )
    results.append(("Small: 10x10", success))
    
    success, *_ = run_test_case(
        "Small divergent (20x20)",
        "A" * 20,
        "C" * 20,
        workers=2
    )
    results.append(("Small: 20x20 divergent", success))
    
    # Medium cases
    print("\n" + "="*60)
    print("MEDIUM CASES")
    print("="*60)
    
    success, t_seq, t_par, speedup = run_test_case(
        "Medium (100x100)",
        generate_sequence(100, seed=10),
        generate_sequence(100, seed=20),
        workers=4
    )
    results.append(("Medium: 100x100", success))
    
    success, t_seq, t_par, speedup = run_test_case(
        "Medium (200x200)",
        generate_sequence(200, seed=30),
        generate_sequence(200, seed=40),
        workers=4
    )
    results.append(("Medium: 200x200", success))
    
    # Large cases (performance check)
    print("\n" + "="*60)
    print("LARGE CASES (Performance Check)")
    print("="*60)
    
    perf_results = []
    
    success, t_seq, t_par, speedup = run_test_case(
        "Large (400x400)",
        generate_sequence(400, seed=100),
        generate_sequence(400, seed=200),
        workers=None  # Use all cores
    )
    results.append(("Large: 400x400", success))
    if speedup is not None:
        perf_results.append(("400x400", t_seq, t_par, speedup))
    
    success, t_seq, t_par, speedup = run_test_case(
        "Large (600x600)",
        generate_sequence(600, seed=300),
        generate_sequence(600, seed=400),
        workers=None
    )
    results.append(("Large: 600x600", success))
    if speedup is not None:
        perf_results.append(("600x600", t_seq, t_par, speedup))
    
    # Summary
    print("\n" + "="*60)
    print("TEST SUMMARY")
    print("="*60)
    
    passed = sum(1 for _, success in results if success)
    total = len(results)
    
    print(f"\nResults: {passed}/{total} tests passed")
    for name, success in results:
        status = "✅ PASS" if success else "❌ FAIL"
        print(f"  {status}: {name}")
    
    if perf_results:
        print(f"\nPerformance Summary:")
        for case, t_seq, t_par, speedup in perf_results:
            print(f"  {case}: {speedup:.2f}x speedup (seq={t_seq:.3f}s, par={t_par:.3f}s)")
    
    # Exit code
    if passed == total:
        print("\n✅ All tests passed!")
        return 0
    else:
        print(f"\n❌ {total - passed} test(s) failed")
        return 1


if __name__ == "__main__":
    sys.exit(main())

File: llm_written/claude-sonnet-4-5/smith-waterman/rust/test_smith_waterman.rs
use sha2::{Sha256, Digest};
use std::time::Instant;
use std::fmt::Write as FmtWrite;

mod algo_sequential;
mod algo_parallel;

use algo_sequential::SmithWaterman as SeqSW;
use algo_parallel::SmithWaterman as ParSW;

fn hash_result(result: &algo_parallel::AlignmentResult) -> String {
    let mut hasher = Sha256::new();
    hasher.update(result.aligned_a.as_bytes());
    hasher.update(result.aligned_b.as_bytes());
    hasher.update(&result.score.to_le_bytes());
    hasher.update(&result.identity.to_le_bytes());
    format!("{:x}", hasher.finalize())
}

fn hash_result_seq(result: &algo_sequential::AlignmentResult) -> String {
    let mut hasher = Sha256::new();
    hasher.update(result.aligned_a.as_bytes());
    hasher.update(result.aligned_b.as_bytes());
    hasher.update(&result.score.to_le_bytes());
    hasher.update(&result.identity.to_le_bytes());
    format!("{:x}", hasher.finalize())
}

fn generate_sequence(len: usize, seed: u64) -> String {
    let bases = ['A', 'C', 'G', 'T'];
    let mut rng = seed;
    (0..len)
        .map(|_| {
            rng = rng.wrapping_mul(6364136223846793005).wrapping_add(1);
            bases[(rng % 4) as usize]
        })
        .collect()
}

struct TestCase {
    name: &'static str,
    query: String,
    reference: String,
}

fn run_test(test: &TestCase, match_score: i32, mismatch_score: i32, gap_score: i32) -> Result<(), String> {
    println!("\n=== Test: {} ===", test.name);
    println!("Query length: {}, Reference length: {}", test.query.len(), test.reference.len());
    
    // Sequential
    let seq_sw = SeqSW::new(match_score, mismatch_score, gap_score);
    let seq_start = Instant::now();
    let seq_result = seq_sw.find_alignment(&test.query, &test.reference);
    let seq_duration = seq_start.elapsed();
    let seq_hash = hash_result_seq(&seq_result);
    
    println!("Sequential:");
    println!("  Time: {:.6}s", seq_duration.as_secs_f64());
    println!("  Score: {}, Identity: {:.2}%", seq_result.score, seq_result.identity);
    println!("  Hash: {}", seq_hash);
    
    // Parallel Run 1
    let par_sw = ParSW::new(match_score, mismatch_score, gap_score);
    let par_start = Instant::now();
    let par_result1 = par_sw.find_alignment(&test.query, &test.reference);
    let par_duration1 = par_start.elapsed();
    let par_hash1 = hash_result(&par_result1);
    
    println!("Parallel Run 1:");
    println!("  Time: {:.6}s", par_duration1.as_secs_f64());
    println!("  Score: {}, Identity: {:.2}%", par_result1.score, par_result1.identity);
    println!("  Hash: {}", par_hash1);
    
    // Parallel Run 2 (determinism check)
    let par_start2 = Instant::now();
    let par_result2 = par_sw.find_alignment(&test.query, &test.reference);
    let par_duration2 = par_start2.elapsed();
    let par_hash2 = hash_result(&par_result2);
    
    println!("Parallel Run 2:");
    println!("  Time: {:.6}s", par_duration2.as_secs_f64());
    println!("  Score: {}, Identity: {:.2}%", par_result2.score, par_result2.identity);
    println!("  Hash: {}", par_hash2);
    
    // Correctness check
    if seq_result.aligned_a != par_result1.aligned_a 
        || seq_result.aligned_b != par_result1.aligned_b
        || seq_result.score != par_result1.score {
        return Err(format!("FAIL: Output mismatch between sequential and parallel"));
    }
    
    // Determinism check
    if par_hash1 != par_hash2 {
        return Err(format!("FAIL: Parallel runs produced different results (hash mismatch)"));
    }
    
    // Speedup
    let speedup = seq_duration.as_secs_f64() / par_duration1.as_secs_f64();
    println!("  Speedup: {:.2}x", speedup);
    
    println!("PASS ✓");
    Ok(())
}

fn main() {
    println!("Smith-Waterman Parallel Differential Test Suite");
    println!("================================================\n");
    
    let match_score = 2;
    let mismatch_score = -1;
    let gap_score = -1;
    
    let mut tests = Vec::new();
    
    // Edge case: empty
    tests.push(TestCase {
        name: "Edge: Empty sequences",
        query: String::new(),
        reference: String::new(),
    });
    
    // Edge case: single char
    tests.push(TestCase {
        name: "Edge: Single character",
        query: "A".to_string(),
        reference: "A".to_string(),
    });
    
    // Edge case: no match
    tests.push(TestCase {
        name: "Edge: No match",
        query: "AAAA".to_string(),
        reference: "TTTT".to_string(),
    });
    
    // Small
    tests.push(TestCase {
        name: "Small: 20x20",
        query: "ACGTACGTACGTACGTACGT".to_string(),
        reference: "ACGTACGTACGTACGTACGT".to_string(),
    });
    
    // Medium
    tests.push(TestCase {
        name: "Medium: 100x100",
        query: generate_sequence(100, 12345),
        reference: generate_sequence(100, 54321),
    });
    
    // Medium-Large
    tests.push(TestCase {
        name: "Medium-Large: 500x500",
        query: generate_sequence(500, 99999),
        reference: generate_sequence(500, 11111),
    });
    
    // Large
    tests.push(TestCase {
        name: "Large: 1000x1000",
        query: generate_sequence(1000, 777777),
        reference: generate_sequence(1000, 888888),
    });
    
    let mut passed = 0;
    let mut failed = 0;
    let mut failure_messages = Vec::new();
    
    for test in &tests {
        match run_test(test, match_score, mismatch_score, gap_score) {
            Ok(_) => passed += 1,
            Err(e) => {
                failed += 1;
                failure_messages.push(format!("{}: {}", test.name, e));
            }
        }
    }
    
    println!("\n================================================");
    println!("Test Summary: {} passed, {} failed", passed, failed);
    
    if !failure_messages.is_empty() {
        println!("\nFailures:");
        for msg in &failure_messages {
            println!("  - {}", msg);
        }
        std::process::exit(1);
    } else {
        println!("All tests passed! ✓");
        std::process::exit(0);
    }
}
